bindings/experimental/distrdf/python/DistRDF/Backends/Spark/Backend.py

```diff
@@ -161,3 +168,39 @@ def make_dataframe(self, *args, **kwargs):
         npartitions = kwargs.pop("npartitions", self.optimize_npartitions())
         headnode = HeadNode.get_headnode(npartitions, *args)
         return DataFrame.RDataFrame(headnode, self)
+
+    @staticmethod
+    def RunGraphs(proxies, numthreads=4):
+        """
+        Trigger multiple RDF graphs through multithreading, according to Spark
+        docs on job scheduling at https://spark.apache.org/docs/latest/job-scheduling.html#scheduling-within-an-application.
```

Comment:
I wonder how well this is going to work on the Spark side. I was reading the Spark docs about this, and it says:
```
By default, Spark’s scheduler runs jobs in FIFO fashion. Each job is divided into “stages” (e.g. map and reduce 
phases), and the first job gets priority on all available resources while its stages have tasks to launch, then the second 
job gets priority, etc. If the jobs at the head of the queue don’t need to use the whole cluster, later jobs can start to run 
right away, but if the jobs at the head of the queue are large, then later jobs may be delayed significantly.
```
In practice this means that `RunGraphs` in Spark only parallelizes if individual jobs don't use all the resources already, otherwise the are just sequential anyway and will run one after the other, pretty much as if we didn't use `RunGraphs`. With the default scheduler, parallelization only happens if e.g. each job uses less partitions than worker processes.

I wonder if using the "fair" scheduler mode (in case we have control over the `SparkContext`) would help in better parallelization even if npartitions>num_workers (e.g. to parallelize the reduce phase of job1 with the map phase of job2).

Reply:
As always with Spark and distributed schedulers, it's not that simple

> RunGraphs in Spark only parallelizes if individual jobs don't use all the resources already, otherwise the are just sequential anyway

If a job takes all available cores in Spark, they will not be released until that job is finished. But, and this is crucial to RunGraphs, as soon as any core is done with Job1's tasks, Job2's tasks that are already enqueued can steal the resources right away. Without RunGraph, there is an overhead in waiting that the full Job1 is over, the Job2 is started (including our startup time in DistRDF which is non-negligible in some cases) and only then Job2 will actually start.

> I wonder if using the "fair" scheduler mode (in case we have control over the SparkContext)

The fair scheduler is not enabled by the SparkContext, it needs to be enabled server-side, i.e. not something we can control from ROOT. If indeed a cluster admin has enabled fair scheduling, then it is only all the more important to provide RunGraphs, so that actually a single user can make full use of the fair scheduling by submitting all jobs at the same time, giving Spark a better chance at guessing how many resources to assign each job. If jobs are submitted sequentially the fair scheduling won't even be used since Job1 will already be over when Job2 is submitted

Reply:
> as soon as any core is done with Job1's tasks, Job2's tasks that are already enqueued can steal the resources right away.

You might be right, but I'd like to measure this. I.e. running the same app with `RunGraphs` and without, with Spark. This sentence `the first job gets priority on all available resources while its stages have tasks to launch` makes me fear that while the reduce stage is ongoing in job1, no tasks for job2 are scheduled.

> The fair scheduler is not enabled by the SparkContext, it needs to be enabled server-side

Did you try this? It's not what seems to be described in the docs: `To enable the fair scheduler, simply set the spark.scheduler.mode property to FAIR when configuring a SparkContext`

Reply:
> Did you try this?

Yes, it's not easy. At the very minimum, different jobs need to belong to different pools as explained [here](https://spark.apache.org/docs/latest/job-scheduling.html#default-behavior-of-pools):
```
but inside each pool, jobs run in FIFO order
```
Then again, creating a different pool means writing at least a `fairscheduler.xml` file and submitting that as a property to the SparkContext. Also this clashes with potentially already available cluster-side `fairscheduler.xml` files or similar ways to define job scheduling and overriding those could lead to unclear scheduling policies.

I have personally tried this by configuring a `YARN` cluster with FAIR scheduling enabled. When it works, it works well. But it still needs jobs to be submitted concurrently. If I have a code like
```python
job1 = sc.parallelize(...).map(...).reduce(...)
job2 = sc.parallelize(...).map(...).reduce(...)
```
It will still run job1 and job2 sequentially, irrespective of any FAIR scheduling enabled on the cluster or by the user. The only way to have them running concurrently and thus actually making use of the FAIR scheduling is by submitting to multiple threads. 

I found this article that also mentions this behaviour and I think it is quite clear https://towardsdatascience.com/apache-spark-sharing-fairly-between-concurrent-jobs-d1caba6e77c2

Reply:
Ok, so we'll need to live with the FIFO scheduler then. With FIFO, I'd like to test if the reducer phase of job1 (when resources are potentially freed) gets interleaved with the mapper phase of job2.